source/slang/slang-file-system.cpp

```diff
@@ -143,45 +156,95 @@ CacheFileSystem::~CacheFileSystem()
-        // Get the path as a string
-        canonicalPath = StringUtil::getString(canonicalBlob);
-    }
     else
     {
-        canonicalPath = m_useSimplifyForCanonicalPath ? Path::Simplify(relPath.getUnownedSlice()) : relPath;
+        // Create and add to canonical path
+        PathInfo* pathInfo  = new PathInfo(canonicalPath);
+        m_canonicalMap.Add(canonicalPath, pathInfo);
+        return pathInfo;
     }
+}
 
-    PathInfo* pathInfo;
-    infoPtr = m_canonicalMap.TryGetValue(canonicalPath);
+CacheFileSystem::PathInfo* CacheFileSystem::_getPathInfo(const String& relPath)
+{
+    PathInfo** infoPtr = m_pathMap.TryGetValue(relPath);
     if (infoPtr)
     {
-        pathInfo = *infoPtr;
+        return *infoPtr;
     }
-    else
+
+    PathInfo* pathInfo = nullptr;
+    switch (m_canonicalMode)
     {
-        // Create and add to canonical path
-        pathInfo = new PathInfo(canonicalPath);
-        m_canonicalMap.Add(canonicalPath, pathInfo);
+        case CanonicalMode::FileSystemExt:
+        {
+            // Try getting the canonical path
+            // Okay request from the underlying file system the canonical path
+            ComPtr<ISlangBlob> canonicalBlob;
+            if (SLANG_FAILED(m_fileSystemExt->getCanoncialPath(relPath.Buffer(), canonicalBlob.writeRef())))
+            {
+                // Write in result as being null ptr so not tried again
+                m_pathMap.Add(relPath, nullptr);
+                return nullptr;
+            }
+            // Get the path as a string
+            String canonicalPath = StringUtil::getString(canonicalBlob);
+            pathInfo = _getPathInfoFromCanonical(canonicalPath);
+            break;
+        }
+        case CanonicalMode::Path:
+        {
+            pathInfo = _getPathInfoFromCanonical(relPath);
+            break;
+        }
+        case CanonicalMode::SimplifiedPath:
+        {
+            pathInfo = _getPathInfoFromCanonical(Path::Simplify(relPath.getUnownedSlice()));
+            break;
+        }
+        case CanonicalMode::Hash:
+        {
+            // I can only see if this is the same file as already loaded by loading the file and doing a hash
+            ComPtr<ISlangBlob> fileBlob;
+            Result res = m_fileSystem->loadFile(relPath.Buffer(), fileBlob.writeRef());
+            if (SLANG_FAILED(res) || fileBlob == nullptr)
+            {
+                // Write in result as being null ptr so not tried again
+                m_pathMap.Add(relPath, nullptr);
+                return nullptr;
+            }
+
+            // Calculate the hash on the contents
+            const uint64_t hash = GetHashCode64((const char*)fileBlob->getBufferPointer(), fileBlob->getBufferSize());
+
+            String hashString = Path::GetFileName(relPath);
+            hashString = hashString.ToLower();
+
+            hashString.append(':');
+
+            // The canonical name is.. combination of name and hash
+            hashString.append(hash, 16);
```

Comment:
Hmm... the phrasing here is a bit odd. We are treating the hash as the canonical name? I would have thought the hash would be a behind-the-scenes thing in the Slang implementation of `#pragma once` rather than part of the logical file system abstraction.

To me this is *not* just a different way of computing canonical paths, and even if the user plugged in an extended file system interface where they'd want to do their own canonicalization we'd still want to have this same basic implementation of `#pragma once`.

Reply:
As it stands 'canonicalPath' can mean a canonical path, or anything else you like - like a hash or whatever, and is documented as such. It is not displayed. This is all based on the previous discussions around this. You could argue it's not a great name - and you'd have a point. You may want to display a 'canonical path' - but that's not the way things are currently set up, as by design (for the above reason) it's never output.